framework/core/js/src/forum/components/UserSecurityPage.tsx

```diff
@@ -0,0 +1,190 @@
+import app from '../../forum/app';
+import UserPage, { IUserPageAttrs } from './UserPage';
+import ItemList from '../../common/utils/ItemList';
+import FieldSet from '../../common/components/FieldSet';
+import listItems from '../../common/helpers/listItems';
+import type Mithril from 'mithril';
+import extractText from '../../common/utils/extractText';
+import AccessTokensList from './AccessTokensList';
+import AccessToken from '../../common/models/AccessToken';
+import LoadingIndicator from '../../common/components/LoadingIndicator';
+import Button from '../../common/components/Button';
+import NewAccessTokenModal from './NewAccessTokenModal';
+import camelCaseToSnakeCase from '../../common/utils/camelCaseToSnakeCase';
+
+/**
+ * The `UserSecurityPage` component displays the user's security control panel, in
+ * the context of their user profile.
+ */
+export default class UserSecurityPage<CustomAttrs extends IUserPageAttrs = IUserPageAttrs> extends UserPage<CustomAttrs> {
+  protected tokens: AccessToken[] | null = null;
+  protected loading: boolean = false;
+
+  oninit(vnode: Mithril.Vnode<CustomAttrs, this>) {
+    super.oninit(vnode);
+
+    const routeUsername = m.route.param('username');
+
+    if (routeUsername !== app.session.user?.slug() && !app.forum.attribute<boolean>('canModerateAccessTokens')) {
+      m.route.set('/');
+    }
+
+    this.loadUser(routeUsername);
+
+    app.setTitle(extractText(app.translator.trans('core.forum.security.title')));
+
+    this.loadTokens();
+  }
+
+  content() {
+    return (
+      <div className="UserSecurityPage">
+        <ul>{listItems(this.settingsItems().toArray())}</ul>
+      </div>
+    );
+  }
+
+  /**
+   * Build an item list for the user's settings controls.
+   */
+  settingsItems() {
+    const items = new ItemList<Mithril.Children>();
+
+    ['developerTokens', 'sessions'].forEach((section) => {
+      const sectionName = `${section}Items` as 'developerTokensItems' | 'sessionsItems';
+      const sectionLocale = camelCaseToSnakeCase(section);
+
+      items.add(
+        section,
+        <FieldSet className={`Security-${section}`} label={app.translator.trans(`core.forum.security.${sectionLocale}_heading`)}>
+          {this[sectionName]().toArray()}
+        </FieldSet>
+      );
+    });
+
+    return items;
+  }
+
+  /**
+   * Build an item list for the user's access accessToken settings.
+   */
+  developerTokensItems() {
+    const items = new ItemList<Mithril.Children>();
+
+    items.add(
+      'accessTokenList',
+      this.tokens === null ? (
+        <LoadingIndicator />
+      ) : (
+        <AccessTokensList
+          key={this.tokens.length}
```

Comment:
Do we need this key?

Reply:
mithril doesn't redraw the list when the tokens change (revoke/create). By using a key with the array length it can tell that it needs to be redrawn.

Not sure why this is necessary here though, I might be missing something about this.

Reply:
I'm not sure either... I can't see any reason why the key would be needed here.

I don't think the key being the array length is a good idea, though, since if (for some reason) the number of tokens stayed the same, but one of them had changed to another model or was edited in some way, we wouldn't trigger a redraw.

I can't think of a better option which doesn't have the same limitation, though.

Reply:
fixed: https://github.com/flarum/framework/pull/3587/commits/61ceb8c8c84f204e5ceb7127244ea427e05e26a8